### ISSUE TITLE
Number type value arguments falsly claim to not over overlap

### ISSUE THREAD

**Issue by PinkFlufflyLlama** [microsoft/TypeScript#49247]
# Bug Report

### 🔎 Search Terms

Type Value Comparison Type Domain Overlap TS2367

### 🕗 Version & Regression Information

- This is an incorrect compiler error
- This behaviour was observed in every version I tried

### ⏯ Playground Link

[Link to the example code](https://www.typescriptlang.org/play?ts=4.6.4#code/C4TwDgpgBAKgrmANhAPDANFAMhAdgc2AAsoIAPYPAEwGcpc4BbAIwgCcA+KAXigG0MUAHQiYfALrioAMigBvKMgLEAXNjyESAXwCwAKFCRYAQ2bI0mAOoBLKsVIVqdBi3aYAEhGv4iwB5VxaeiZWTh5YBHN4JFRBGzsiDg8vH2AOfX0AY0RjGjoAWWNgNmsyFHj7cgCgl1Dk719-J2DXMLl9KE6oTIB7XBpiuEzgHrYACjA4M2tMqCoi4zUYU3NatygKonrUjgBKeV09Dq78CD9rGgBlAEc4YzYIMf32vS63qAfgODZcKGILoTzYDGIRKTQ8bi8f40QELPgABnEoI0xGOnUOWiAA)

### 💻 Code

```ts
type Tuple<T, Length extends number> = [T, ...T[]] & { length: Length }
type Table<T, Width extends number, Height extends number> = Tuple<Tuple<T, Width>, Height>

class Matrix<Width extends number, Height extends number> {
    constructor(public data: Table<number, Width, Height>) {}

    get isSquare() {
        return this.data.length === this.data[0].length
    }
}
```

### 🙁 Actual behaviour

Causes error `TS2367: This condition will always return 'false' since the types 'number & Height' and 'number & Width' have no overlap.`
There seems to be several issues here:
- Why is number & Height not absorbed into just Height? - Might not be a bug
- Why is it number & Height in the first place? - Might not be a bug
- Why is there an error occurring here? Both Width and Height extend number and thus there should be ample overlap between the two types. - Defiantly an issue.

My best guess at what is occurring is the type checker doesn't have the capability to look at the "parent" of Width and Height and notice that they both have the same "parent"

[Casting the types](https://www.typescriptlang.org/play?ts=4.6.4#code/C4TwDgpgBAKgrmANhAPDANFAMhAdgc2AAsoIAPYPAEwGcpc4BbAIwgCcA+KAXigG0MUAHQiYfALrioAMigBvKMgLEAXNjyESAXwCwAKFCRYAQ2bI0mAOoBLKsVIVqdBi3aYAEhGv4iwB5VxaeiZWTh5YBHN4JFRBGzsiDg8vH2AOfX0AY0RjGjoAWWNgNmsyFHj7cgCgl1Dk719-J2DXMLl9KE6oTIB7XBpiuEzgHrYACjA4M2tMqCoi4zUYU3NatygKonrUjgBKeV09Dq78CD9rGgBlAEc4YzYIMf32vS63qAfgODZcKDHiC5CebAYxCJSaKC5KCeBrAfbcBF-AE0IELPgABnEYI09ihm12x06hy0QA) to Width and Height and doing the comparison still has the domain error.

### 🙂 Expected behaviour

No errors occur

**Comment by jcalz** [comment]
Related to #47026; comparisons between two generic types are not usually allowed.

The wording of the error message "This condition will always return `false` because the types Foo and Bar don't overlap" isn't great, in my opinion.  I'd prefer something like "Types Foo and Bar are not directly related to each other and comparing them is probably a mistake; if this is intentional, please assert them to a common type."  I guess that message is probably going to stay, though...  #26592 is open, but #36534 is closed as "won't fix"

**Comment by PinkFlufflyLlama** [comment]
Its certainly disappointing that any checking doesn't occur, but I guess the complexity of the type checker needs to be considered, but it certainly should have a better compiler message needs to be used because it is a valid statement.

**Comment by RyanCavanaugh** [comment]
I think as part of #26592 we need to remove the "will always return false" verbiage
